Replace pci_enable_msix() with pci_alloc_irq_vectors()

andreas-rollbuehler opened the pull request:
Since kernel version 4.8. pci_enable_msix() isn't available any more.
Therefore it is replaced with pci_alloc_irq_vectors() if used
kernel version is >= 4.8

Signed-off-by: Andreas Rollbühler <andreas.rollbuehler@siemens.com>

henning-schild commented on kernel_module/uio/uio_ivshmem.c:
```diff
@@ -15,6 +16,8 @@
 #include <linux/uio_driver.h>
 #include <linux/io.h>
 
+#include <linux/version.h>
```
please add this right under the other includes, some people like to order them somehow but i do not care

henning-schild commented on kernel_module/uio/uio_ivshmem.c:
```diff
@@ -101,12 +111,27 @@ static int request_msix_vectors(struct ivshmem_info *ivs_info, int nvectors)
+#else
+	err = pci_alloc_irq_vectors(ivs_info->dev, 1, ivs_info->nvectors,
+			PCI_IRQ_MSIX);
+
```
Indentation of the second line is wrong

henning-schild commented on kernel_module/uio/uio_ivshmem.c:
```diff
@@ -85,6 +93,8 @@ static int request_msix_vectors(struct ivshmem_info *ivs_info, int nvectors)
 		ivs_info->msix_entries[i].entry = i;
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4,8,0)
+
```
skip the empty line

henning-schild commented on kernel_module/uio/uio_ivshmem.c:
```diff
@@ -101,12 +111,27 @@ static int request_msix_vectors(struct ivshmem_info *ivs_info, int nvectors)
+	} else {
+		ivs_info->nvectors = err;
+	}
+#endif
```
I think with a bit of restructuring both the error handling and the assignment of "nvectors = err" can be reused across the two cases.

henning-schild commented on kernel_module/uio/uio_ivshmem.c:
```diff
@@ -101,12 +111,27 @@ static int request_msix_vectors(struct ivshmem_info *ivs_info, int nvectors)
 			"%s-config", name);
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4,8,0)
+
```
skip empty line

henning-schild commented on kernel_module/uio/uio_ivshmem.c:
```diff
@@ -115,14 +140,26 @@ static int request_msix_vectors(struct ivshmem_info *ivs_info, int nvectors)
+			goto error;
+		}
+#endif
 	}
```
The only real change is the first argument for request_irq, right? If you factor that out all the rest can be reused in both cases instead of copying it.

Did the return value of request_irq change? Why the change from if (err) to if (err < 0)?

henning-schild commented on kernel_module/uio/uio_ivshmem.c:
```diff
@@ -115,14 +140,26 @@ static int request_msix_vectors(struct ivshmem_info *ivs_info, int nvectors)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,8,0)
+	pci_free_irq_vectors(ivs_info->dev);
+#endif
 	return err;
```
Something similar will also have to go into ivshmem_pci_remove. I think that should be called when you unbind the device from the driver or when you rmmod the module.

henning-schild reviewed (changes requested):
Thanks, i found a few minor style problems and it seems the free is missing on one path.

andreas-rollbuehler committed:
Refactor code for subsequent patch

This patch doesn't change any functionality. It just refactors some code to make
the changes in the subsequent patch easier to read.

Signed-off-by: Andreas Rollbühler <andreas.rollbuehler@siemens.com>

andreas-rollbuehler committed:
Replace pci_enable_msix() with pci_alloc_irq_vectors()

Since kernel version 4.8. pci_enable_msix() isn't available any more.
Therefore it is replaced with pci_alloc_irq_vectors() if the used
kernel version is >= 4.8

Signed-off-by: Andreas Rollbühler <andreas.rollbuehler@siemens.com>
